Add feature to auto create users for locations

esoergel opened the pull request:
@orangejenny
https://docs.google.com/document/d/18m4HhuFL5ALdtbHnN5nvn234J_SynP1o5jPgHFn_jXQ/edit#

esoergel committed:
Add location-user model fields

esoergel committed:
Add UI to update location_type.has_user

esoergel committed:
Make meta location form for the various entities

This form edits both location, custom location data, and soon user and user
data, so I figure it makes sense to give them each their own form and tie 'em
together with a meta form

esoergel committed:
Add user and custom_user_data forms to locations page

esoergel committed:
Hide the user forms if not applicable

esoergel committed:
Automatically create user

esoergel committed:
cross-link location and user

esoergel committed:
Check for uniqueness on save, not just client-side

esoergel committed:
Stick location users behind a feature flag

esoergel committed:
Fire form signals appropriately

esoergel committed:
Link user and location archival/deletion

esoergel committed:
create or archive users on loctypes change

esoergel committed:
Add user_location_id to ES index

esoergel committed:
Catch easy error

esoergel committed:
Misc fixes

esoergel committed:
Test this bad boy out

dimagimon commented on corehq/apps/locations/forms.py:
```diff
@@ -337,6 +304,153 @@ def save(self, instance=None, commit=True):
+    """Ties together the forms for location, location data, user, and user data."""
+
+    def __init__(self, location, request_user, is_new, bound_data=None, *args, **kwargs):
+        from .views import LocationFieldsView
```
F401 'LocationFieldsView' imported but unused

dimagimon commented on corehq/apps/locations/tasks.py:
```diff
@@ -37,3 +44,81 @@ def sync_supply_points(location_type):
+    username = format_username(with_suffix, domain)
+    if not CommCareUser.username_exists(username):
+        return username
+    return _get_unique_username(domain, base, suffix+1, tries_left-1)
```
E226 missing whitespace around arithmetic operator

esoergel committed:
lint [ci skip]

sravfeyn commented:
Hey can you guys review this @orangejenny @biyeun Ethan is offline, so [I am going to address](https://trello.com/c/J5R3DPkN/110-virtual-location-users) PR feedback to get it in

orangejenny commented on corehq/apps/locations/forms.py:
```diff
@@ -305,17 +273,17 @@ def clean_coordinates(self):
         location.site_code = self.cleaned_data['site_code']
         location.location_type = self.cleaned_data['location_type_object']
-        location.metadata = self.custom_data.get_data_to_save()
+        location.metadata = metadata or {}
```
So if `metadata` isn't passed in (as [here](https://github.com/dimagi/commcare-hq/pull/15774/files#diff-1d8a6e2b00dbb7850a4c1bfde042d6e8R523)), `location.metadata` gets cleared out? Making metadata a required parameter would make this clearer.

orangejenny commented on corehq/apps/locations/forms.py:
```diff
@@ -337,6 +304,152 @@ def save(self, instance=None, commit=True):
+            location_data = self.custom_location_data.get_data_to_save()
+            location = self.location_form.save(metadata=location_data)
+            self.user.user_location_id = location.location_id
+            self.user.set_location(location)
```
Would it be reasonable to put `self.user.user_location_id = location.location_id` inside of `set_location`?

esoergel replied:
not quite, since `user_location_id` is a different thing - users can be assigned to locations (many-to-one), but these are location-users (one-to-one), which are much more closely coupled.  I had trouble coming up with names that made this distinction clear, and would welcome suggestions if you've got 'em.

orangejenny commented on corehq/apps/locations/forms.py:
```diff
@@ -337,6 +304,152 @@ def save(self, instance=None, commit=True):
+            self.domain,
+            username,
+            password,
+            device_id="Generated from HQ",
```
Is it worth being more specific about why this was generated?

esoergel replied:
Actually I don't know what this is used for, or if it is used at all.  I just copied that form the NewMobileWorkerView.

orangejenny commented on corehq/apps/users/forms.py:
```diff
@@ -537,8 +537,9 @@ def __init__(self, project, user, *args, **kwargs):
         email_string = u"@{}.commcarehq.org".format(project.name)
         max_chars_username = 80 - len(email_string)
         self.project = project
+        self.domain = self.project.name
```
Not a big deal, but curious what the purpose of this change was.

esoergel replied:
I think it was just that `self.domain` was available on the other form classes here, and I was mildly annoyed when it didn't work here.

orangejenny commented on corehq/apps/users/views/mobile/users.py:
```diff
@@ -844,7 +855,7 @@ def _build_commcare_user(self):
         return CommCareUser.create(
             self.domain,
-            format_username(username, self.domain),
+            username,
```
Can you clarify what this change was for? Looks like it could have far-reaching side effects.

esoergel replied:
For sure.  `format_username` isn't gone - I added `clean_mobile_worker_username` to the `clean_username` method, and that function includes the functionality of `format_username`.  Actually when I first added that, I was getting users like `ethan@test.commcarehq.org@test.commcarehq`, until I realized I needed to drop this call.  I added that clean function because we were only validating username uniqueness on the client-side.

orangejenny replied:
aha, got it, thanks for the explanation!

orangejenny commented on corehq/apps/users/views/mobile/users.py:
```diff
@@ -405,6 +405,12 @@ def delete_commcare_user(request, domain, user_id):
+    if (user.user_location_id and
+            SQLLocation.objects.get_or_None(location_id=user.user_location_id,
+                                            user_id=user._id)):
+        messages.error(request, "This is a location user. You must delete the "
```
translate

mkangia replied:
@esoergel this needs a change right?

orangejenny commented:
Looks reasonable. The only possibly blocking comment is getting clarification on what `format_username` was removed from `_build_commcare_user`.

esoergel committed:
Fix for strong passwords

esoergel committed:
Make custom data prefix configurable

esoergel committed:
Send signals a little more gracefully

If there are failures, still send the signal so handlers can add additional errors

esoergel committed:
Pass only location_type code, that's easier

esoergel committed:
Don't attempt to process usertype when it's missing

If it's missing, the custom user data validation will have caught it

esoergel committed:
Pass custom_data to the signal

dimagimon commented on custom/enikshay/tests/test_user_setup.py:
```diff
@@ -57,7 +57,11 @@ def tearDownClass(cls):
     def assertValid(self, form):
-        msg = "{} has errors: \n{}".format(form.__class__.__name__, form.errors.as_text())
+        if isinstance(form, LocationFormSet):
+            errors = ", ".join([form.errors.as_text() for form in form.forms])
```
F812 list comprehension redefines 'form' from line 59

esoergel committed:
List comprehensions should reaaaally have limited scope

[ci skip]

esoergel committed:
make toggling has_user and strong password easier

esoergel committed:
wip

esoergel commented:
@mkangia would you mind doing a review of this?  I added some changes since the last one.

esoergel commented:
or @sravfeyn

mkangia commented on corehq/apps/locations/migrations/0009_add_location_user_fields.py:
```diff
@@ -0,0 +1,25 @@
+        migrations.AddField(
+            model_name='sqllocation',
+            name='user_id',
+            field=models.CharField(blank=True, max_length=255),
```
Do we need an index for this?

mkangia commented:
@esoergel Could not deep dive into this but just did a browse through. left a suggestion on migration done. Defer to @sravfeyn or @orangejenny to review again for changes.

orangejenny commented:
New commits look fine to me, up to you if you want to merge or get feedback from @sravfeyn

esoergel committed:
translate

esoergel committed:
Make metadata required

esoergel committed:
Clarify logic regarding including user forms

esoergel committed:
Forgot to reset the user ids on archive

esoergel committed:
Handle locations changing type to or from has_user

esoergel committed:
Whoopsie - forgot to update this test.

esoergel commented:
This is awaiting QA

esoergel committed:
Don't fail on repr(unsaved_location)

esoergel committed:
Revert "wip"

This reverts commit 906ecd229b6a29e33365f6af964c8747b74d7585.

esoergel committed:
Infer location_type if there's only one option
